### PR TITLE
MSVC Support

### PR DESCRIPTION
**Pull request by carsten-grimm-at-ipolog**
This branch includes several changes to help add support for MSVC to lager. 
* When using MSVC, lager now requires C++20. This fixes several issues related to template resolution.
* the following MSVC-specific compiler flags were added:
  * /Wall to enable all warnings,  
  * /Zc:preprocessor to enable variadic macros, and
  * /bigobj to allow msvc to handle larger object files, which is required to handle situations with many inline functions (e.g., the test `xformed` fails without this flag) .
 * Currently, MSVC cannot handle the compile-time dependency checks when creating a store in lager. Hence, the build option `lager_DISABLE_STORE_DEPENDENCY_CHECKS` was added to disable these checks. The option is off by default to retain the current behavior for non-MSVC compilers, but it is forced on when using MSVC.
 * Some libraries may not be available when developing with MSVC. The requirements for these libraries were loosened:
   * Examples using `ncurses` are only build, when `cmake` can find `ncurses`.
   * The build option `lager_BUILD_DEBUGGER` was added to toggle building the lager debugger and all examples that use it. This helps in situations where the dependency on `libhttpserver` cannot be satisfied. The option is on by default to retain the current behavior.
 * Switching to C++20 for MSVC caused deprecation errors in the external library `catch`. The library was updated to version v2.12.2 that supports C++20.
 * Some fixes were required for the tests
   * Some of the core tests were accessing `std:nullopt`, which is UB.
   * Some reference to reference errors were fixed (which may have been introduced by the upgrade of `catch`).
   * Some tests use the counter example. The inclusion of the corresponding header no longer depends on the build paths.
   * The test targets now have the include directory for boost.
   * The lenses test was missing an include for `std::array` (according to MSVC).
 * For the sake of convenience, the configuration folders for Visual Studio were added to `.gitignore`.

There are some points to note:
* These changes were tested with both `MSVC 19.30.30706.0` and `Clang 12.0.0 with MSVC-like command-line` . 
* `immer` was used (via a modified `vcpkg`) using the build from commit `42e6beafed53f2ecd971360270f421f9c2e36642`.
* `zug` was used (via a modified `vcpkg`) using the build from commit `033dadbed463ff3430b7ebc36df8a46f4f0f5078`.
* None of `ncurses`, `libhttpserver`, `sdl`, and `imgui` were available for testing. 

There is one remaining issue that I was unable to fix: The test `test-event_loop-queue` is **failing** with a read access error. It appears that a lambda capture failed, leading to a call on an uninitialized object. This occurs with MSVC, but it does not occur with Clang (neither for C++17 nor for C++20).

This happens to be my first pull request on github. Suggestions for improvements are welcome.

### CODE REVIEW

File: lager/deps.hpp
Comment (arximboldi) [inline lager/deps.hpp:347]: Nitpick: I may slightly prefer this form:
```c++
    template <
        typename... Ds
#ifndef LAGER_DISABLE_STORE_DEPENDENCY_CHECKS
       , std::enable_if_t<required_key_set == boost::hana::intersection(
                                                 required_key_set,
                                                 deps<Ds...>::required_key_set),
                         bool> = true>
            required_key_set,
            deps<Ds...>::required_key_set),
        bool> = true
#endif
    >
```
Same below.

**Review by arximboldi — COMMENTED** [review]
This is great, than you so much for this effort Carsten! Let me know if you want to address the nitpicks, otherwise I will merge once the tests are green and I have a chance to test locally (I'm on a terribly slow connection on a bus right now :smile: )

**Comment by carsten-grimm-at-ipolog** [comment]
> This is great, than you so much for this effort Carsten! Let me know if you want to address the nitpicks, otherwise I will merge once the tests are green and I have a chance to test locally (I'm on a terribly slow connection on a bus right now 😄 )

Thanks for the feedback. I will implement your suggestions tomorrow. I might try out the Windows Subsystem for Linux to see if I can get the tests to pass on my machine before resubmitting.

**Comment by carsten-grimm-at-ipolog** [comment]
Hi @arximboldi. 

Here is a summary of the changes from the update to this pull request:
* Disabling the compile-time store dependencies was simplified as requested.
* The build option for the examples using the debugger has been renamed to `lager_BUILD_DEBUGGER_EXAMPLES`. The corresponding build target has been renamed to `lager-debugging-example` and is now stated as a specialization of the build target `lager-example`. I have reordered the CMakeLists.txt a little bit to keep everything related to examples in one place. It has to be placed below the section for the tests, since the build target for the checks is defined in that section.
* The examples are only added the `checks` target, if tests are enabled. This fixes an issue where cmake would fail with examples enabled and tests disabled.
* Some details of the tests `test-xformed` and `test-deps` were changed in order to please gcc: 
  * Duplicate `const` was removed from certain exception related checks in `test-xformed` and `test-deps`. 
  * The dependency on `boost::fusion` was removed from `test-xformed` as it seemed incompatible with catch; the comparison in question was implemented using `std::tie` instead.
  * An include for `optional` was added to `test-xformed`.

I have tested this with MSVC, Clang, and with gcc via the github action.

Let me know if there is anything amiss or if you are not a fan of some change.

File: lager/deps.hpp
Comment (carsten-grimm-at-ipolog) [reply lager/deps.hpp]: This is fixed.

File: CMakeLists.txt
Comment (carsten-grimm-at-ipolog) [reply CMakeLists.txt]: The requested change is implemented.

File: CMakeLists.txt
Comment (carsten-grimm-at-ipolog) [reply CMakeLists.txt:36]: Thanks :-)

**Review by arximboldi — CHANGES_REQUESTED** [review]
Hi Carsten! This is great and I really appreciate the division in meaningful steps in commits and attention to detail. There is just one small mistake, you forgot to rename `lager_BUILD_DEBUGGER` to `lager_BUILD_DEBUGGER_EXAMPLES` in `examples/CMakeLists.txt`, which causes some examples to not be built anymore. Can you fix that before I merge? Otherwise, local tests are also running fine and I will be happy to merge once that is fixed. Thank you again so much for this!

**Comment by arximboldi** [comment]
Oh, sorry, another issue:
```
if(Curses)
```
Should be:
```
if(CURSES_FOUND)
```

**Comment by carsten-grimm-at-ipolog** [comment]
Hi @arximboldi ! 

Thanks you for the feedback and sorry for the two issues. They should be fixed now: The check for `ncurses` uses `Courses_FOUND` and all occurrences of `lager_BUILD_DEBUGGER` have been replaced by `lager_BUILD_DEBUGGER_EXAMPLES`.

**Pull request merged**
